hq_superset/views.py

```diff
@@ -0,0 +1,162 @@
+from zipfile import ZipFile
+from .utils import (get_datasource_export_url, get_ucr_database, get_schema_name_for_domain,
+    get_datasource_list_url)
+from .oauth import get_valid_cchq_oauth_token
+from .hq_domain import user_domains
+
+
+class HQDatasourceView(BaseView):
+
+    def __init__(self):
+        self.route_base = "/hq_datasource/"
+        super().__init__()
+
+    @expose("/update/<datasource_id>", methods=["GET"])
+    def create_or_update(self, datasource_id):
+        # Fetches data for a datasource from HQ and creates/updates a superset table
+        from .oauth import get_valid_cchq_oauth_token
+        res = refresh_hq_datasource(g.hq_domain, datasource_id)
+        return res
+
+    @expose("/list/", methods=["GET"])
+    def list_hq_datasources(self):
+        datasource_list_url = get_datasource_list_url(g.hq_domain)
+        provider = superset.appbuilder.sm.oauth_remotes["commcare"]
+        oauth_token = get_valid_cchq_oauth_token()
+        response = provider.get(datasource_list_url, token=oauth_token)
+        return self.render_template(
+            "hq_datasource_list.html",
+            datasources=response.json(),
+        )
+
+
+
+class CCHQApiException(Exception):
+    pass
+
+
+def refresh_hq_datasource(domain, datasource_id):
+    # This method pulls the data from CommCareHQ and creates/replaces the
+    #   corresponding Superset dataset
+    datasource_url = get_datasource_export_url(domain, datasource_id)
+    provider = superset.appbuilder.sm.oauth_remotes["commcare"]
+    oauth_token = get_valid_cchq_oauth_token()
+    response = provider.get(datasource_url, token=oauth_token)
+    if response.status_code != 200:
+        # Todo; logging
+        raise CCHQApiException("Error downloading the UCR export from HQ")
+    zipfile = ZipFile(BytesIO(response.content))
+    filename = zipfile.namelist()[0]
+    # Upload to table
+    database = get_ucr_database()
+    schema = get_schema_name_for_domain(domain)
+    csv_table = Table(table=datasource_id, schema=schema)
+
+    try:
+        df = pd.concat(
+            pd.read_csv(
+                chunksize=1000,
+                filepath_or_buffer=zipfile.open(filename),
+                encoding="utf-8",
+                # Todo; make date parsing work
+                parse_dates=True,
+                infer_datetime_format=True,
+                keep_default_na=True,
+            )
+        )
+
+        database.db_engine_spec.df_to_sql(
+            database,
+            csv_table,
+            df,
+            to_sql_kwargs={
+                "chunksize": 1000,
+                "if_exists": "replace",
+            },
+        )
+
+        # Connect table to the database that should be used for exploration.
+        # E.g. if hive was used to upload a csv, presto will be a better option
+        # to explore the table.
+        expore_database = database
+        explore_database_id = database.explore_database_id
```

Comment:
How is this id getting populated?

Reply:
https://github.com/dimagi/hq_superset/pull/7/files#r870067051